### ISSUE TITLE
Git LFS hook

### ISSUE THREAD

**Issue by detouched** [brigade/overcommit#477]
Looks like it's not possible to use an arbitrary command in PATH in a custom hook, although it should be, according to the docs.

From documentation:
> Name of an executable that must exist in order for the hook to run. If this is a path (e.g. ./bin/ruby), ensures that the executable file exists at the given location relative to the repository root. **Otherwise, if it just the name of an executable (e.g. ruby) checks if the executable can be found in one of the directories in the PATH environment variable.** Set this to a specific path if you want to always use an executable that is stored in your repository. (e.g. RubyGems bin stubs, Node.js binaries, etc.)

However, my attempts to use custom hook that calls some command in PATH results in the following signature verification error:
```
Hook must specify a `required_executable` or `command` that is tracked by git (i.e. is a path relative to the root of the repository) so that it can be signed
```

In particular, the setup for Git LFS mentioned in [this comment](https://github.com/brigade/overcommit/pull/369#issuecomment-212193259) doesn't work now.
```yml
PrePush:
  GitLfs:
    enabled: true
    command: ['git-lfs', 'pre-push']
```

**Comment by sds** [comment]
Hey @detouched,

Understandable why this is confusing. Unfortunately the example you're referring to is incorrect&mdash;you need to specify the `required_executable` option instead of `command`. You can see an example in the [documentation](https://github.com/brigade/overcommit#adding-existing-git-hooks).

I've edited the example in the other thread to use `required_executable` instead so hopefully no one else gets thrown off by this.

Hope this helps!

**Comment by detouched** [comment]
Hm, unfortunately that doesn't work either. Trying with this:

```
$ cat .overcommit.yml
PrePush:
  GitLfs:
    enabled: true
    required_executable: 'git-lfs'
    flags: ['pre-push']

$ git push
Hook must specify a `required_executable` or `command` that is tracked by git (i.e. is a path relative to the root of the repository) so that it can be signed
/usr/local/lib/ruby/gems/2.3.0/gems/overcommit-0.38.0/lib/overcommit/hook_signer.rb:39:in `hook_path'
...
```

git-lfs binary is available in PATH

**Comment by sds** [comment]
Sorry, I misunderstood the situation. 

["Ad hoc" hooks](https://github.com/brigade/overcommit#adding-existing-git-hooks) require the hook to be tracked by source control so that Overcommit can "sign" the executable, for reasons discussed in the [Security](https://github.com/brigade/overcommit#security) section.

An easy way around this in your case is to create a shim script in your repository, say `script/git-lfs` so you can refer to it and have it be tracked by Git.

```bash
# script/git-lfs
#!/bin/bash
exec git-lfs "$@"
```

Is this silly? Perhaps in this case, but it's a protection to ensure you understand that Overcommit can't sign executables that aren't under the domain of the Git repo it watches.

Hope that makes some sense.

**Comment by detouched** [comment]
Yeah, I understand the reason for signing the hooks. So what that sentence from `required_executable` means then?
> Otherwise, if it just the name of an executable (e.g. ruby) checks if the executable can be found in one of the directories in the PATH environment variable.

Or is it about the case when the hook needs an executable but is not that executable itself?

Well, for LFS it indeed would be counter-intuitive to commit a hook in the repository. As a workaround, I'm going to put it into `.git/hooks/custom` and redefine `plugin_directory`, same as I did with my other custom plugin. And copy it into the template directory so it appears in new repositories automatically.

Thanks for your help!

**Comment by sds** [comment]
It may be the case that this "security" isn't helpful, and perhaps should be removed. I would definitely be open to that, if there wasn't a nagging feeling in the back of my mind that we had a good reason for the existing behavior. However, if the reason was particularly good, it should have been documented.

Appreciate your thoughts on this! Sorry for the rough edges around this particular feature.

**Comment by detouched** [comment]
Well, it's a tough question. Since overcommit might execute some code from the repository, I believe it's generally a good idea to protect the user by forcing him to confirm the changes. On the other hand, it's quite easy to work around part of the signing, as you mentioned: by creating a shim script that calls something else. The shim script will be signed, while that other script won't, and it might be harmless. However, even notifying about change in `.overcommit.yml` (by forcing user to sign it) is a good feature I think.

To summarize, I think signing is a nice feature so far, and it should stay.

As for the LFS case, what if we add its pre-push hook into bundled hooks? It'll be signed similar to, say, [SpellCheck](https://github.com/brigade/overcommit/blob/master/lib/overcommit/hook/commit_msg/spell_check.rb) which also calls some command from PATH. I'd also add an example to uncomment into `default.yml` in addition to the docs.

If that sounds good, I'll raise a PR...

**Comment by sds** [comment]
Adding a hook for Git LFS sounds like a great idea! Happy to accept a PR.
